File: java/L100/LeetCode0106.java
package L100;

import util.TreeNode;

public class LeetCode0106 {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        return build(inorder, 0, inorder.length - 1, postorder, 0, postorder.length - 1);
    }
    private TreeNode build(int[] inorder,int il,int ir,int[] postorder,int pl,int pr) {
        if (il > ir || pl > pr) {
            return null;
        } else if (il == ir) {
            return new TreeNode(inorder[il]);
        }
        TreeNode node = new TreeNode(postorder[pr]);
        int ri = il;
        for (int i = il;i <= ir;++i) {
            if (inorder[i] == postorder[pr]) {
                ri = i;
                break;
            }
        }
        int pi = pl + ri - il - 1;
        node.left = build(inorder, il, ri - 1, postorder, pl, pi);
        node.right = build(inorder, ri + 1, ir, postorder, pi + 1, pr - 1);
        return node;
    }
}

File: cpp/L100/LeetCode0106.cc
#include "../heads/tree_node.h"
#include <vector>
using std::vector;

class Solution {
    int findIndex(vector<int> &inorder, int ilow, int iHigh, int value) {
        for (int i = ilow;i <= iHigh;++i) {
            if (inorder[i] == value) {
                return i;
            }
        }
        return -1;
    }
    TreeNode* buildChildTree(vector<int> &inorder, int ilow, int iHigh, vector<int> &postorder, int plow, int phigh) {
        if (ilow > iHigh || plow > phigh) {
            return nullptr;
        }
        TreeNode *node = new TreeNode(postorder[phigh]);
        int id = findIndex(inorder, ilow, iHigh, postorder[phigh]);
        int lz = id - ilow, rz = iHigh - id;
        node->left = buildChildTree(inorder, ilow, id - 1, postorder, plow, plow + lz - 1);
        node->right = buildChildTree(inorder, id + 1, iHigh, postorder, plow + lz, phigh - 1);
        return node;
    }
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        int iz = inorder.size(), pz = postorder.size();
        return buildChildTree(inorder, 0, iz - 1, postorder, 0, pz - 1);
    }
};